source/simulator/assembly.cc

```diff
@@ -1419,7 +1419,8 @@ namespace aspect
                    * on the direction of the flow. Thus the flow-dependent terms below are
                    * only calculated if the edge is an inflow edge.
                    */
-                  const bool inflow = ((current_u * scratch.face_finite_element_values->normal_vector(q)) < 0.);
+                  const bool inflow = (advection_field.is_temperature()
+                                       || ((current_u * scratch.face_finite_element_values->normal_vector(q)) < 0.));
```

Comment:
Hm, I think this requires a bit more study. I definitely see that there is a question that needs to be answered. Are you willing to remove this part of the patch and put it into a separate pull request? The other parts are uncontroversial and should be merged; they are also logically independent of this issue, and we try to keep logically separate issues in separate pull requests.